ipmininet/ipnet.py

```diff
@@ -549,6 +550,79 @@ def ping6Pair(self):
            return: ploss packet loss percentage"""
         return self.pingPair(use_v4=False)
 
+   
+    def runFailurePlan(self, failure_plan: List[Tuple[str,str]]) -> List[IPIntf]:
+        """this function run a failure plan
+            :param: A list of tuple of string giving the name of the 2 nodes between
+                    which the link have to be taken down
+            :return: A list of IPIntf which represent the interface that goes down or 
+                    an empty list if there was an unknown node in the failure_plan
+        """
+        print("** Starting failure plan")
+        interfaces_down = []
+        for link in failure_plan:
+            try:
+                node1 = self.get(link[0])
+                node2 = self.get(link[1])
+                interfaces = node1.connectionsTo(node2)
+                interfaceNode1 = interfaces[0][0]
+                interfaceNode2 = interfaces[0][1]
+                interfaces_down.append(interfaceNode1)
+            except KeyError as e:
+                log.output("ERROR: node "+ str(e) +"\n")
+                interfaces_down = []
+        for interface in interfaces_down:
+            commande = "ip link set dev " + str(interface) + " down"
+            interface.node.cmd(commande)
+            print("** Interface "+ str(interface) +" down")
+        return interfaces_down
+
+    def restoreLink(self, interfaces: List[IPIntf]):
+        """function which restore the link
+            :param interfaces: a List of IPIntf
+        """
+        print("** starting restoring link")
+        for interface in interfaces:
+            commande = "ip link set dev " + str(interface) + " up"
+            interface.node.cmd(commande)
+            print("** interfaces " + str(interface) + " up")
+
+    def RandomFailure(self, n, weak_links=None) -> List[IPIntf]:
+        """ this function down randomly n link
+            :param n: the number of link to be downed
+            :param weak_links: an optional parameter which specify a list of IPIntf in
+                                which the n links will be downed
+            :return: a list of IPIntf which have been downed
+        """
+        if weak_links is None:
+            all_links = self.links
```

Comment:
I was actually thinking to something much simpler than what you implemented:

```suggestion
            all_links = weak_links if weak_links is not None else self.links
```

The idea is just to restrict the links on which random runs, not to shutdown all the links in a method called "RandomFailure". It would be strange

Reply:
I'm not sure to understand the issue here. not all the link are downed here, either no restricted link are specified (the weak_links which is a list of IPIntf) and then only n links of all the links are downed using self.links(which is a List of IPLink). Or a subset of links to down is specify using the weak_links parameter and then only n links are downed in this list.
The issue why this implementation seems so complicated is because self.links return a list of IPLinks while weak_links are a list of IPIntf. Because I've not found in the doc how we can retrieve a list of IPLink based on a node.

Reply:
> Because I've not found in the doc how we can retrieve a list of IPLink based on a node.

You can use ``intf.link`` to get the parent link of an interface

> The issue why this implementation seems so complicated is because self.links return a list of IPLinks while weak_links are a list of IPIntf.

To be consistent, ``weak_links`` should be a list of ``IPLink`` then
And this should simplify your implementation